asm/callonstack.s

```diff
@@ -8,13 +8,25 @@
 //     [RSI] long (*func)(void* arg),
 //     [RDX] void* arg);
 //
-// Call a function on the given stack
+// Call a function on the given stack.
 //
+// GDB normally will stop stack-walking at a function if it determines that
+// the stack does not monotonically decrease at the function.
+// This could be the case for some invocations of myst_call_on_stack, if the
+// `stack` parameter is greater than RSP upon entry.
+// To support the split-stack feature, GDB however relaxes the above
+// constraint for functions named `__morestack`.
+// We leverage this GDB behavior by moving the body of myst_call_on_stack to
+// a local function name `__morestack` to which myst_call_on_stack jumps to.
 //==============================================================================
-
 .globl myst_call_on_stack
 .type myst_call_on_stack, @function
 myst_call_on_stack:
+   jmp __morestack
```

Comment:
how does gdb make this work? The return address on the stack is myst_call_on_stack still, not __morestack

Reply:
When__morestack calls bar, the return address that gets pushed on the stack belongs to __morestack. Thus GDB split stack backtrace works.

However, since myst_call_on_stack jumps to __morestack instead of calling it, myst_call_on_stack does not appear in the backtrace.